pkg/config/remote/service/service.go

```diff
@@ -1167,6 +1294,161 @@ func (s *CoreAgentService) ConfigResetState() (*pbgo.ResetStateConfigResponse, e
 	return &pbgo.ResetStateConfigResponse{}, err
 }
 
+// CreateConfigSubscription creates a new config subscription for a client.
+//
+// The client can send requests to track or untrack products for clients with a
+// given ID. If a client is tracked, all files for products that the
+// subscription tracks should be streamed to this subscription when that client
+// would receive those files via the regular ClientGetConfigs RPC.
+func (s *CoreAgentService) CreateConfigSubscription(
+	stream pbgo.AgentSecure_CreateConfigSubscriptionServer,
+) error {
+	// Register this subscription, checking for limits on total subscriptions.
+	subID, updateSignal, err := func() (subscriptionID, <-chan struct{}, error) {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		if len(s.mu.subscriptions.subs) >= s.maxConcurrentSubscriptions {
+			return 0, nil, status.Errorf(
+				codes.ResourceExhausted,
+				"maximum number of subscriptions reached (%d)",
+				s.maxConcurrentSubscriptions,
+			)
+		}
+		subID, updateSignal := s.mu.subscriptions.newSubscription()
+		s.telemetryReporter.SetConfigSubscriptionsActive(len(s.mu.subscriptions.subs))
+		return subID, updateSignal, nil
+	}()
+	if err != nil {
+		log.Warnf("failed to create subscription: %v", err)
+		return err
+	}
+	defer func() {
+		s.telemetryReporter.IncConfigSubscriptionsDisconnectedCounter()
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		s.mu.subscriptions.remove(subID)
+		s.telemetryReporter.SetConfigSubscriptionsActive(len(s.mu.subscriptions.subs))
+	}()
+	s.telemetryReporter.IncConfigSubscriptionsConnectedCounter()
+	// Send a header to synchronize with the client and prevent client-side
+	// retries.
+	if err := stream.SendHeader(metadata.MD{}); err != nil {
+		return err
+	}
+	popUpdate := func() *pbgo.ConfigSubscriptionResponse {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		return s.mu.subscriptions.popUpdate(subID)
+	}
+	setTrackedClientsLocked := func() {
+		n := 0
+		for _, s := range s.mu.subscriptions.subs {
+			n += len(s.trackedClients)
+		}
+		s.telemetryReporter.SetConfigSubscriptionClientsTracked(n)
+	}
+	track := func(runtimeID string, products pbgo.ConfigSubscriptionProducts) error {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		sub := s.mu.subscriptions.subs[subID]
+		if len(sub.trackedClients) >= s.maxTrackedRuntimeIDsPerSubscription {
+			return status.Errorf(
+				codes.ResourceExhausted,
+				"maximum number of runtime IDs per subscription reached (%d)",
+				s.maxTrackedRuntimeIDsPerSubscription,
+			)
+		}
+		sub.track(runtimeID, products)
+		setTrackedClientsLocked()
+		return nil
+	}
+	untrack := func(runtimeID string) {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		s.mu.subscriptions.subs[subID].untrack(runtimeID)
+		setTrackedClientsLocked()
+	}
```

Comment:
Nit: These aren't used by your anon goroutine that processes updates, can they be moved closed down the the loop processing the gRPC stream messages where they are used?